### ISSUE TITLE
ValidationPipe only works in Gateways if validateCustomDecorators=true

### ISSUE THREAD

**Issue by da-mkay** [nestjs/nest#3127]
## Current behavior
Using below code I try to use the `ValidationPipe` to validate against my `AuthenticateDto` class.

1. Without using the `@MessageBody` decorator, the transform-method of `ValdiationPipe` is never called. Is that correct? It is not mentioned in the [docs](https://docs.nestjs.com/websockets/pipes).
2. But even if `@MessageBody` is used, I have to add `{validateCustomDecorators: true}` to the pipe's options. Otherwise the `ValidationPipe` skips validation.

## Input Code

```ts
@UsePipes(new ValidationPipe())
@SubscribeMessage('authenticate')
authenticate(
    @ConnectedSocket() client: Socket,
    @MessageBody() data: AuthenticateDto
) {
    // ...
}
```

## Expected behavior
It should not be neccessary to pass `validateCustomDecorators: true` to get validation working, because we also don't have to do it in HTTP controller context.
Maybe it should also not be neccessary to use the `@MessageBody` decorator or the docs are wrong :)

## Possible Solution
Inside the `ValidationPipe` ([here](https://github.com/nestjs/nest/blob/99bd8676afa48efd56661c35aadd534627bf22ce/packages/common/pipes/validation.pipe.ts#L108-L110)) it is checked whether the (data-)argument is of type `"custom"` and if so whether `validateCustomDecorators` is true. The question here should be: why is the type `"custom"` and not, e.g., `"body"`? I used the `@MessageBody` decorator, so it should be some kind of body/payload-type right?
The `"custom"`-string is created [here](https://github.com/nestjs/nest/blob/99bd8676afa48efd56661c35aadd534627bf22ce/packages/core/pipes/params-token-factory.ts#L5,L15) inside `ParamsTokenFactory`, based on the `type`-parameter passed in which is of type `RouteParamtypes`.
Well, in this case, we use websockets, so the `type`-parameter is actually of type [`WsParamtype`](https://github.com/nestjs/nest/blob/master/packages/websockets/enums/ws-paramtype.enum.ts). Okay, but how could it get there?
Check out this [line](https://github.com/nestjs/nest/blob/master/packages/core/pipes/pipes-consumer.ts#L14) of `PipesConsumer`. The type-castings allow any value to be passed there. 
To sum up, `WsContextCreator` uses that `PipesConsumer` which uses `ParamsTokenFactory` which, as described earlier, considers only `RouteParamtypes`, not `WsParamtype` to determine the type-string. Thus, for websockets the type is always `"custom"`.
This looks kinda wrong to me. Shouldn't the `PipesConsumer` work with different XyzTokenFactory's, depending on the context? `ParamsTokenFactory` for HTTP and some new `WsTokenFactory`for websockets?

## Environment

<pre><code>
Nest version: 6.8.2
 
For Tooling issues:
- Node version: 10.16.3  
- Platform: Mac 
</code></pre>

Maybe this issue is related: #3114

**Comment by baflo** [comment]
I had a similar problem. Without digging too deep, I'd say it's the same. In my case it was introduced with `6.8.0`. 

In `@nestjs/websockets` there's a newly defined constant `PARAM_ARGS_METADATA` which seems to never being used to set metadata. 

However, it's tried to be used at several points, e.g. in [`ws-context-creator.ts#L152`](https://github.com/nestjs/nest/blob/44bc0edd87bffed02440235a68e8573b8bf6a8e1/packages/websockets/context/ws-context-creator.ts#L152). The metadata that should be defined for `PARAM_ARGS_METADATA` is stored to `metadata`, which is in turn used by `getParamsMetadata`, that always returns an empty array `[]` as `metadata` is always empty (as it's never written).

Consequently, `paramsOptions` ([`ws-context-creator.ts#L82`](https://github.com/nestjs/nest/blob/44bc0edd87bffed02440235a68e8573b8bf6a8e1/packages/websockets/context/ws-context-creator.ts#L81)) is always an empty array, and eventually `pipesFn` is never returned ([`ws-context-creator.ts#L235`](https://github.com/nestjs/nest/blob/44bc0edd87bffed02440235a68e8573b8bf6a8e1/packages/websockets/context/ws-context-creator.ts#L237)).

This causes the `ValidationPipe` never being invoked.

**Comment by kamilmysliwiec** [comment]
Fixed in 6.8.3. Please, let me know if you face any issue.

**Comment by da-mkay** [comment]
Awesome, 6.8.3 works (no validateCustomDecorators or MessageBody needed anymore). Thank you!
